### PR TITLE
Additional ingress rules for control plane

### PR DESCRIPTION
**Pull request by alexander-demicev**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This PR introduces the ability to specify additional ingress rules for the control plane.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/392

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Additional ingress rules for the control plane can be specified in the API
```

### CODE REVIEW

**Comment by richardcase** [comment]
@alexander-demicev - the MD remediation test failure isn't related to this change, its generally failing.

**Review by richardcase — COMMENTED** [review]
Thanks @alexander-demicev :bow:  This is looking good to me. A couple of very minor points:

- Might be nice to add coverage of this to the e2e tests
- Do you think it would be worth adding something to the docs?
- Could you squash the commits?

**Comment by richardcase** [comment]
Also, just checking the EKS side as there is some shared code here:

/test pull-cluster-api-provider-aws-e2e-eks

**Review by mjura — APPROVED** [review]
CI test are failing, but it looks good to me

File: api/v1beta2/awscluster_webhook.go [hunk 1]
```diff
@@ -190,3 +191,19 @@ func (r *AWSCluster) validateNetwork() field.ErrorList {
...
+func (r *AWSCluster) validateAdditionalIngressRules() field.ErrorList {
+	var allErrs field.ErrorList
+
+	if r.Spec.ControlPlaneLoadBalancer == nil {
+		return allErrs
+	}
+
+	for _, rule := range r.Spec.ControlPlaneLoadBalancer.AdditionalIngressRules {
+		if (rule.CidrBlocks != nil || rule.IPv6CidrBlocks != nil) && (rule.SourceSecurityGroupIDs != nil || rule.SourceSecurityGroupRoles != nil) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("additionalIngressRules"), r.Spec.ControlPlaneLoadBalancer.AdditionalIngressRules, "CIDR blocks and security group IDs cannot be used together"))
```
Comment (furkatgofurov7) [inline api/v1beta2/awscluster_webhook.go:204]: It is not only `SourceSecurityGroupIDs` as per the condition, so probably should be something like:
```suggestion
			allErrs = append(allErrs, field.Invalid(field.NewPath("additionalIngressRules"), r.Spec.ControlPlaneLoadBalancer.AdditionalIngressRules, "CIDR blocks and security group IDs or security group roles cannot be used together"))
```

File: api/v1beta2/awscluster_webhook_test.go [hunk 2]
```diff
@@ -251,6 +251,107 @@ func TestAWSClusterValidateCreate(t *testing.T) {
...
+								Protocol:               SecurityGroupProtocolTCP,
+								IPv6CidrBlocks:         []string{"test"},
+								SourceSecurityGroupIDs: []string{"test"},
+							},
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
```
Comment (furkatgofurov7) [inline api/v1beta2/awscluster_webhook_test.go:287]: this is a duplicate of previous test case on lines 254-270

**Review by furkatgofurov7 — COMMENTED** [review]
Few nits, overall looks good

**Comment by richardcase** [comment]
@alexander-demicev - it would be good to have an e2e test to cover this new feature. We could do it as part of this PR or as a follow-up. wdyt?

Apart from that i think this is ready to go, great work :+1:

**Comment by alexander-demicev** [comment]
@richardcase done

**Comment by richardcase** [comment]
Lets do the e2e as follow-up

/lgtm

**Comment by richardcase** [comment]
For final approval: @Skarlso @Ankitasw

File: api/v1beta2/network_types.go [hunk 3]
```diff
@@ -530,10 +531,15 @@ var (
...
 // IngressRule defines an AWS ingress rule for security groups.
 type IngressRule struct {
-	Description string                `json:"description"`
-	Protocol    SecurityGroupProtocol `json:"protocol"`
-	FromPort    int64                 `json:"fromPort"`
-	ToPort      int64                 `json:"toPort"`
+	// Description is a description of the ingress rule.
+	Description string `json:"description"`
+	// Protocol is the protocol for the ingress rule. Accepted values are "-1" (all), "4" (IP in IP),"tcp", "udp", "icmp", and "58" (ICMPv6).
+	// +kubebuilder:validation:Enum="-1";"4";tcp;udp;icmp;"58"
```
Comment (Skarlso) [inline api/v1beta2/network_types.go:537]: Add the capitalized versions as well like in TargetGroup spec.

File: api/v1beta2/network_types.go [hunk 4]
```diff
@@ -530,10 +531,15 @@ var (
 
 // IngressRule defines an AWS ingress rule for security groups.
 type IngressRule struct {
-	Description string                `json:"description"`
-	Protocol    SecurityGroupProtocol `json:"protocol"`
-	FromPort    int64                 `json:"fromPort"`
-	ToPort      int64                 `json:"toPort"`
+	// Description is a description of the ingress rule.
```
Comment (Skarlso) [inline api/v1beta2/network_types.go:534]:
```suggestion
	// Description provides extended information about the ingress rule.
```

File: pkg/cloud/services/securitygroup/securitygroups.go [hunk 5]
```diff
@@ -519,6 +519,20 @@ func (s *Service) getSecurityGroupIngressRules(role infrav1.SecurityGroupRole) (
 		if s.scope.Bastion().Enabled {
 			rules = append(rules, s.defaultSSHIngressRule(s.scope.SecurityGroups()[infrav1.SecurityGroupBastion].ID))
 		}
+		if s.scope.ControlPlaneLoadBalancer() != nil {
```
[the remark this answers is not in the document]
Comment (Ankitasw) [reply pkg/cloud/services/securitygroup/securitygroups.go:522]: nvm, we can take it as a follow up. I just saw @richardcase 's comment 🙂 
overall LGTM.

**Pull request merged**